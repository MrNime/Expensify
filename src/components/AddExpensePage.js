import React from 'react';

const AddExpensePage = () => <div>This is from add expense</div>;

export default AddExpensePage;
