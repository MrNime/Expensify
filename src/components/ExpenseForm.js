import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import { addExpense } from '../actions/expenses';

export default class ExpenseForm extends Component {
  state = {
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    calenderFocused: false,
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    const regex = /^\d{1,}(\.\d{0,2})?$/;
    if (!amount || amount.match(regex)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onCalendarFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };
  render() {
    return (
      <div>
        <form>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onCalendarFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            displayFormat="D/MM/YYYY"
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
