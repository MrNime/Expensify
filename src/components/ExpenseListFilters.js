import React, { Component } from 'react';
import { connect } from 'react-redux';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';

import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
  clearFilters,
} from '../actions/filters';

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onCalendarFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    // eslint-disable-next-line
    e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
  };
  onClearFilters = () => {
    this.props.clearFilters();
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              placeholder="Search expenses"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onCalendarFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              displayFormat="DD/MM/YYYY"
            />
          </div>
          <div className="input-group__item">
            <button
              type="button"
              aria-label="Clear Dates"
              className="DateRangePickerInput_clearDates DateRangePickerInput_clearDates_1 clear-button"
              onClick={this.onClearFilters}
            >
              <svg
                className="DateRangePickerInput_clearDates_svg DateRangePickerInput_clearDates_svg_1"
                viewBox="0 0 12 12"
              >
                <path
                  fillRule="evenodd"
                  d="M11.53.47a.75.75 0 0 0-1.061 0l-4.47 4.47L1.529.47A.75.75 0 1 0 .468 1.531l4.47 4.47-4.47 4.47a.75.75 0 1 0 1.061 1.061l4.47-4.47 4.47 4.47a.75.75 0 1 0 1.061-1.061l-4.47-4.47 4.47-4.47a.75.75 0 0 0 0-1.061z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters,
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  clearFilters: () => dispatch(clearFilters()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
