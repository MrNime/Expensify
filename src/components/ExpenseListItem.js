import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({
  id, description, amount, paidAmount, createdAt,
}) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div className="list-item__summary">
      <h3 className="list-item__title">{description}</h3>
      <span className="list-item__subtitle">{moment(createdAt).format('Do MMMM YYYY')}</span>
    </div>
    <div className="list-item__amounts">
      <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
      <h3 className="list-item__data list-item__data--small">
        {numeral(paidAmount / 100).format('$0,0.00')}
      </h3>
      <h3 className="list-item__data list-item__data--small">
        {numeral((amount - paidAmount) / 100).format('$0,0.00')}
      </h3>
    </div>
  </Link>
);

export default ExpenseListItem;
