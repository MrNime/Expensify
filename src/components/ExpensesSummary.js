import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal, { savedTotal } from '../selectors/expenses-total';

export const ExpensesSummary = ({
  visibleExpenseCount,
  visibleExpenseTotal,
  visibleExpenseSavings,
  invisibleExpenseCount,
  invisibleExpenseTotal,
  allExpenseCount,
  allExpenseTotal,
  allExpenseSavings,
}) => {
  const visibleExpenseWord = visibleExpenseCount === 1 ? 'expense' : 'expenses';
  const formattedVisibleExpenseTotal = numeral(visibleExpenseTotal / 100).format('$0,0.00');
  const formattedVisibleExpenseSavings = numeral(visibleExpenseSavings / 100).format('$0,0.00');
  const invisibleExpenseWord = invisibleExpenseCount === 1 ? 'expense' : 'expenses';
  const formattedInvisibleExpenseTotal = numeral(invisibleExpenseTotal / 100).format('$0,0.00');
  const allExpenseWord = allExpenseCount === 1 ? 'expense' : 'expenses';
  const formattedAllExpenseTotal = numeral(allExpenseTotal / 100).format('$0,0.00');
  const formattedAllExpenseSavings = numeral(allExpenseSavings / 100).format('$0,0.00');
  return (
    <div className="page-header">
      <div className="content-container">
        <div className="page-header__content">
          <h1 className="page-header__title">
            Viewing <span>{visibleExpenseCount}</span> {visibleExpenseWord} worth{' '}
            <span>{formattedVisibleExpenseTotal}</span>. Visible savings:{' '}
            <span>{formattedVisibleExpenseSavings}</span>
          </h1>
          <div className="page-header__section">
            <Link className="button" to="/create">
              Add Expense
            </Link>
            <h3 className="page-header__subtitle">
              <div className="page-header__filtered">
                Filtering <span>{invisibleExpenseCount}</span> {invisibleExpenseWord} worth{' '}
                <span>{formattedInvisibleExpenseTotal}</span>
              </div>
              <div className="page-header__total">
                from <span>{allExpenseCount}</span> total {allExpenseWord} worth{' '}
                <span>{formattedAllExpenseTotal}</span>. Total saved:{' '}
                <span>{formattedAllExpenseSavings}</span>
              </div>
            </h3>
          </div>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    visibleExpenseCount: visibleExpenses.length,
    visibleExpenseTotal: selectExpensesTotal(visibleExpenses),
    visibleExpenseSavings: savedTotal(visibleExpenses),
    invisibleExpenseCount: state.expenses.length - visibleExpenses.length,
    invisibleExpenseTotal:
      selectExpensesTotal(state.expenses) - selectExpensesTotal(visibleExpenses),
    allExpenseCount: state.expenses.length,
    allExpenseTotal: selectExpensesTotal(state.expenses),
    allExpenseSavings: savedTotal(state.expenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
