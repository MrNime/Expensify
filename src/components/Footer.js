import React from 'react';

const Footer = ({ dispatchStartLogout }) => (
  <header className="footer">
    <div className="content-container">
      <div className="footer__content">
        <p>
          Coded by{' '}
          <a href="https://github.com/NickyMeuleman/Expensify" target="_blank" rel="noopener noreferrer">
            Nicky Meuleman
          </a>
        </p>
        <p>
          Based on a course by{' '}
          <a href="https://mead.io/" target="_blank" rel="noopener noreferrer">
            Andrew J. Mead
          </a>
        </p>
      </div>
    </div>
  </header>
);

export default Footer;
