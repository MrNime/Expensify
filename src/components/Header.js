import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogout } from '../actions/auth';
import ConnectedUserWidget from '../components/UserWidget';

export const Header = ({ dispatchStartLogout, isAuthenticated }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard">
          <h1>Expensify</h1>
        </Link>
        {isAuthenticated && <ConnectedUserWidget />}
      </div>
    </div>
  </header>
);

const mapDispatchToProps = dispatch => ({
  dispatchStartLogout: () => dispatch(startLogout()),
});

const mapStateToProps = state => ({
  isAuthenticated: !!state.auth.uid,
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
