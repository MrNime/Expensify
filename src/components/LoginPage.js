import React from 'react';
import { connect } from 'react-redux';
import { startLogin, startFacebookLogin } from '../actions/auth';

export const LoginPage = ({ dispatchStartLogin, dispatchStartFacebookLogin }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Expensify</h1>
      <p>It&apos;s time to get your expenses under control</p>
      <button className="button box-layout__button" onClick={dispatchStartLogin}>
        Login with Google
      </button>
      <button className="button" onClick={dispatchStartFacebookLogin}>
        Login with Facebook
      </button>
    </div>
  </div>
);

const mapDispatchToProps = dispatch => ({
  dispatchStartLogin: () => dispatch(startLogin()),
  dispatchStartFacebookLogin: () => dispatch(startFacebookLogin()),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
