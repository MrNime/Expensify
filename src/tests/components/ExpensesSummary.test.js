import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should correctly render ExpensesSummary with 0 expenses', () => {
  const wrapper = shallow(<ExpensesSummary
    visibleExpenseCount={0}
    visibleExpenseTotal={0}
    invisibleExpenseCount={0}
    invisibleExpenseTotal={0}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary
    visibleExpenseCount={1}
    visibleExpenseTotal={123}
    invisibleExpenseCount={0}
    invisibleExpenseTotal={0}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary
    visibleExpenseCount={14}
    visibleExpenseTotal={123456789}
    invisibleExpenseCount={0}
    invisibleExpenseTotal={0}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummeray with 1 filtered expense', () => {
  const wrapper = shallow(<ExpensesSummary
    visibleExpenseCount={14}
    visibleExpenseTotal={123456789}
    invisibleExpenseCount={1}
    invisibleExpenseTotal={23894}
  />);
  expect(wrapper).toMatchSnapshot();
})

test('should correctly render ExpensesSummeray with multiple filtered expenses', () => {
  const wrapper = shallow(<ExpensesSummary
    visibleExpenseCount={14}
    visibleExpenseTotal={123456789}
    invisibleExpenseCount={832}
    invisibleExpenseTotal={23894101903}
  />);
  expect(wrapper).toMatchSnapshot();
})

