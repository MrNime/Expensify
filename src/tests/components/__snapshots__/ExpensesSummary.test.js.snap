// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 0 expenses 1`] = `
<div>
  <h1>
    Viewing 
    0
     
    expenses
     totalling 
    € 0,00
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    € 1,23
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    14
     
    expenses
     totalling 
    € 1 234 567,89
  </h1>
</div>
`;
